### PR TITLE
[nexus] put DNS servers in DNS, so you can DNS while you DNS

### LINKED ISSUE #4889: DNS propagation should use internal DNS to find DNS servers
Currently, the DNS propagation stuff uses the `services` table to enumerate the list of DNS servers that it's responsible for keeping up to date.  It could instead use one of these:

- DNS itself.  I realize this sounds absurd.  But internal DNS servers operate at known addresses -- by design -- so that they can always be discovered.  And they have to be up and discoverable for Nexus to even come up and find CockroachDB.  So I can't see a plausible scenario where this approach would fail but one of the other approaches here would work.  To use this, we'd need to be sure we correctly create DNS records for internal and external DNS servers.
- Once blueprints are available, this could use the current target blueprint instead.  That should be about the same as using DNS.  One difference is that systems today won't necessarily have a target blueprint yet so we'd need another code path for those (or else block this task until the target blueprint is established).
- As a fallback, we could use a recent inventory collection.  I think this is probably never the best option, but it might be an okay fallback if we decide to use the target blueprint but there isn't one yet.  But I think this is a reason to prefer using DNS instead.

Importantly, all of these will stay up-to-date as either internal or external DNS servers are added or removed.

### PR DESCRIPTION
**Pull request by hawkw**
Currently, the DNS propagation background task in Nexus uses the `services` table to enumerate the list of DNS servers that it's responsible for keeping up to date.  However, we'd really like to get rid of said `services` table (see #4947), and the DNS propagation code is the only remaining user of the `services` table.

Therefore, this branch changes how DNS servers are discovered for DNS propagation. Rather than discovering DNS server addresses from the `services` table, the `DnsWatcher` background task now discovers DNS servers...using internal DNS. As described in  #4889, this may _seem_ like a cyclical dependency, but, because the initial set of internal DNS servers operate at known addresses -- by design -- so that they can always be discovered. And they have to be up and discoverable for Nexus to even come up and find CockroachDB.  So, internal DNS can safely be assumed to be up if Nexus has come up at all. Now, the `services` table is no longer used, and 

This change breaks the existing tests `nexus::app::background::init::test_dns_propagation_basic` and `nexus::app::background::dns_servers::test_basic`. I've rewritten the `test_dns_propagation_basic` test to test the new expected behavior:
-  creating a new internal DNS server and adding a DNS record for it to the database's DNS config table results in that server's DNS records being propagated to the existing DNS serve
- the `DnsWatcher` background task then picks up the DNS records for the new DNS server by querying the existing known DNS server
- the current DNS config generation is then propagated to the new DNS server
- a subsequent generation is propagated to both the initial and new DNS servers

The `dns_servers::test_basic` test tested the discovery of DNS server addresses from the database. Because these no longer come from the db, and now come from internal DNS, this test would now end up exercising most of the functionality tested in `test_dns_propagation_basic`. I didn't think it was necessary to have two tests for this, so I made the judgement call to delete `dns_servers::test_basic`. If we think having a more isolated test that exercises only  the DNS watcher task and not the DNS propagation task, we could put this back and create records on the DNS server by manually hitting its API with new configs, but I didn't think this was really worth the effort.

I've also removed the `Datastore::upsert_service` method, which was used only for test code and is now dead. I considered deleting all code related to querying the `services` table in this branch as well. However, I noticed that it's still populated when initializing the rack, and that `omdb` has commands for querying that table. I wasn't sure if there were alternative data sources for the `omdb` debugging commands yet, so I didn't remove them. If the data provided by those commands is available elsewhere, or if their only real purpose is _just_ to print the state of this table, I'm happy to delete them in this branch, as well.

Closes #4889

image

### CODE REVIEW

**Comment by smklein** [comment]
> Therefore, this branch changes how DNS servers are discovered for DNS propagation. Rather than discovering DNS server addresses from the `services` table, the `DnsWatcher` background task now discovers DNS servers...using internal DNS. As described in #4889, this may _seem_ like a cyclical dependency, but, because the initial set of internal DNS servers operate at known addresses -- by design -- so that they can always be discovered. And they have to be up and discoverable for Nexus to even come up and find CockroachDB. So, internal DNS can safely be assumed to be up if Nexus has come up at all. Now, the `services` table is no longer used, and

I think this might have gotten cut-off!

> I've also removed the `Datastore::upsert_service` method, which was used only for test code and is now dead. I considered deleting all code related to querying the `services` table in this branch as well. However, I noticed that it's still populated when initializing the rack, and that `omdb` has commands for querying that table. I wasn't sure if there were alternative data sources for the `omdb` debugging commands yet, so I didn't remove them. If the data provided by those commands is available elsewhere, or if their only real purpose is _just_ to print the state of this table, I'm happy to delete them in this branch, as well.

I think it's wise to defer the deletion to a separate branch, just to keep things decoupled. Happy with this direction though.

File: nexus/src/app/background/init.rs [hunk 1]
```diff
@@ -391,6 +389,7 @@ pub mod test {
     //     the new DNS configuration
     #[nexus_test(server = crate::Server)]
     async fn test_dns_propagation_basic(cptestctx: &ControlPlaneTestContext) {
+        const INTERNAL_DNS_SRV_NAME: &'static str = "_nameservice._tcp";
```
[the remark this answers is not in the document]
Comment (hawkw) [reply nexus/src/app/background/init.rs:392]: ah, that's much nicer, thanks!

File: nexus/src/app/background/dns_servers.rs [hunk 2]
```diff
@@ -75,58 +65,38 @@ impl BackgroundTask for DnsServersWatcher {
...
+                                "failed to read list of DNS servers: {:#}",
+                                error
+                            )
+                    });
+                }
+                Ok(addresses) => {
+                    // TODO(eliza): it would be nicer if `Resolver` had a method
+                    // returning an iterator instead of a `Vec`, so we didn't
+                    // have to drain the Vec and then collect it into a new
+                    // one...
```
[the remark this answers is not in the document]
Comment (hawkw) [reply nexus/src/app/background/dns_servers.rs:93]: yeah, i might do that, although i didn't think it was particularly important to minimize allocation overhead in the background task --- this code doesn't seem particularly hot.

[the remark this answers is not in the document]
Comment (luqmana) [reply nexus/src/app/background/dns_servers.rs:93]: The lookup methods on `Resolver` could definitely use a fresh pass. But agreed, it's not a blocker here

File: nexus/src/app/background/init.rs [hunk 3]
```diff
@@ -467,29 +476,76 @@ pub mod test {
...
+            2,
+        )
+        .await;
+
+        let config = dns_config_client
+            .dns_config_get()
             .await
-            .unwrap();
+            .expect("failed to get initial DNS server config");
+        println!("config: {config:#?}");
```
Comment (luqmana) [inline nexus/src/app/background/init.rs:531]: Leftover from testing? The `wait_propagate_dns` call right above should validate getting the initial DNS server's config and that it has the right generation at this point.

Comment (hawkw) [reply nexus/src/app/background/init.rs:531]: ah, yup, good catch --- I can remove this

**Review by luqmana — APPROVED** [review]

**Review by davepacheco — APPROVED** [review]
Thanks for doing this and @luqmana and @smklein for the review.  I appreciate the extra eyes on this.

**Comment by davepacheco** [comment]
In terms of #4987: I agree with not removing all of the `service` table stuff in this PR.  I'd like to avoid doing that until we've finished with its replacement (i.e., until blueprints are far enough along that we've automated their generation so that we can always look at a blueprint when we want to know what zones should exist).  Ideally I'd love to avoid building any _new_ code that uses this table but I can't think of a great way to do that aside from a comment in the datastore file to that effect.

File: nexus/src/app/background/init.rs [hunk 4]
```diff
@@ -467,29 +476,70 @@ pub mod test {
             SocketAddr::V4(_) => panic!("expected v6 address"),
             SocketAddr::V6(a) => a,
         };
+
+        let update = {
```
[the remark this answers is not in the document]
Comment (hawkw) [reply nexus/src/app/background/init.rs:480]: good call, I'll add that!

**Comment by hawkw** [comment]
> In terms of #4987: I agree with not removing all of the `service` table stuff in this PR. I'd like to avoid doing that until we've finished with its replacement (i.e., until blueprints are far enough along that we've automated their generation so that we can always look at a blueprint when we want to know what zones should exist). Ideally I'd love to avoid building any _new_ code that uses this table but I can't think of a great way to do that aside from a comment in the datastore file to that effect.

One option would be removing the `Datastore` wrappers that actually query the table, in order to discourage its use in new code, but continuing to populate it in case we need it, until it's no longer necessary. Of course, we can still write new queries for the table, but not having a method for it should maybe make it less easy to reach for. If you think that this approach makes sense, I've opened #5044 to do that.

**Pull request merged**